clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -3581,8 +3582,10 @@ ConstantAddress CodeGenModule::GetAddrOfTemplateParamObject(
       isExternallyVisible(TPO->getLinkageAndVisibility().getLinkage())
           ? llvm::GlobalValue::LinkOnceODRLinkage
           : llvm::GlobalValue::InternalLinkage;
-  auto *GV = new llvm::GlobalVariable(getModule(), Init->getType(),
-                                      /*isConstant=*/true, Linkage, Init, Name);
+  auto *GV = new llvm::GlobalVariable(
+      getModule(), Init->getType(),
+      /*isConstant=*/true, Linkage, Init, Name, nullptr,
+      llvm::GlobalValue::NotThreadLocal, GlobalsInt8PtrTy->getAddressSpace());
```

Comment:
This doesn't do anything unless the datalayout is broken: globals already go into getDefaultGlobalsAddressSpace() by default

Reply:
At the same time, it's not terribly costly (except for lines of code), and, unfortunately, there might be quite a few "broken" datalayouts (NVPTX is in the same boat, as is DirectX). I'm not entirely convinced the breakage is not an intentional design choice for pseudo/meta/abstract targets (abusing nomenclature, apologies). I'd submit that this (the changes/CodeGen) is still in Clang, which would still fall under the "the FE/language has to deal with ASes having special properties, LLVM shouldn't care", and it is an artifact of composing AS aware languages (OCL/CUDA/HIP etc.) with said abstract targets.

Reply:
If the datalayout is broken, codegen and optimizations break in other ways... the important thing here is consistency.  It isn't really relevant whether the backend ultimately lowers to ELF, or WASM, or SPIRV; everything needs to agree on the properties of different address-spaces.  For example, there are IR optimizations that create globals.

So I'd much rather fix this problem now, rather than continue to scatter workarounds across the codebase.

If the datalayout can't be fixed for some reason, we can discuss that, but I don't see any obvious reason why that would be the case.

Reply:
I've opened #88455 to fix SPIR & SPIRV, which'll allow simplifying this one in the direction you have indicated.

Reply:
This has been merged, so I'm going to mark this as done. Thank you.